File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        StringBuilder str = new StringBuilder(s);
        int n = str.length();
        Stack<Integer> st = new Stack();
        
        for(int i = 0; i < n; i++) {
            char ch = str.charAt(i);
            if(ch == '(')
                st.add(i);
            else if(ch == ')') {
                if(st.size() > 0 && str.charAt(st.peek()) == '(') {
                    st.pop();
                } else {
                    str.setCharAt(i, '*');
                }
            }
        }
        
        while(st.size() != 0) {
            str.setCharAt(st.peek(), '*');
            st.pop();
        }
        
        String ans = "";
        for(int i = 0; i< n; i++) {
            if(str.charAt(i) == '*') {
                continue;
            } else {
                ans += str.charAt(i);
            }
        }
        
        return ans;
    }
}

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        int n = s.size();
        stack<int> st;
        vector<bool> toRemove(n, false);
        
        for(int i = 0; i < n; i++) {
            char ch = s[i];
            if(ch == '(')
                st.push(i);
            else if(ch == ')') {
                if(st.size() > 0 && s[st.top()] == '(') {
                    st.pop();
                } else {
                    toRemove[i] = true;
                }
            }
        }
        
        while(st.size() != 0) {
            toRemove[st.top()] = true;
            st.pop();
        }
        
        string ans = "";
        for(int i = 0; i< n; i++) {
            if(toRemove[i] == true) {
                continue;
            } else {
                ans += s[i];
            }
        }
        
        return ans;
    }
};